Fix CIFS Mount/Unmount

THX723 opened the pull request:
Addressed hangs on initial mount/unmount test, if the mount points are unreachable.
Added timeout to mount and unmount.
Streamlined the codebase.
Only export ARCHIVE_MOUNT and MUSIC_ARCHIVE_MOUNT if using CIFS server.

THX723 committed:
Update archiveloop

Only export ARCHIVE_MOUNT and MUSIC_ARCHIVE_MOUNT if using CIFS server.

THX723 committed:
Update disconnect-archive.sh

Fixed flawed logic stuck on initial test, if mount point is already inaccessible.
Added timeout for umount and streamlined the code.

THX723 committed:
Update archiveloop

Typo

THX723 committed:
Update archiveloop

Typo

THX723 commented:
@marcone ready for review.

marcone commented on run/cifs_archive/disconnect-archive.sh:
```diff
@@ -1,29 +1,26 @@
+      log "$mount_point unmounted."
+    else
+      log "$mount_point failed to unmount!"
     fi
   fi
+}
 
-  if [ -e "$MUSIC_ARCHIVE_MOUNT" ]
-  then
-    log "unmounting $MUSIC_ARCHIVE_MOUNT"
-    if ! umount -f -l "$MUSIC_ARCHIVE_MOUNT"
-    then
-      log "unmount failed"
-    fi
-  fi
+{
+  unmount_if_set "${ARCHIVE_MOUNT:-}"
+  unmount_if_set "${MUSIC_ARCHIVE_MOUNT:-}"
```
This is not entirely correct: /mnt/archive might not exist , because TeslaUSB can be configured without a cam_disk. The unmount will fail in that case, though the failure would be handled gracefully.
This could be improved by having archiveloop only export ARCHIVE_MOUNT and MUSIC_ARCHIVE_MOUNT if those folders actually exist.

THX723 replied:
I did just that in `archiveloop`. Did you miss it?

THX723 replied:
Reread your comment. I'll add a check for /mnt/archive too.

marcone reviewed (commented):
This

THX723 committed:
Update archiveloop

Added check for CIFS cam share as well.
Also simplified the code for the checks.

THX723 commented:
@marcone ok, should be good to go now.

THX723 commented:
@marcone Hold on this. I'm realizing the same issue applies to CIFS `connect-archive.sh` as well. Need to look into it more.

THX723 committed:
Update disconnect-archive.sh

checked mount exists before unmounting.
Rephrase output to match other parts of the archiveloop.

THX723 committed:
Update connect-archive.sh

Fixed flawed logic stuck on initial test, if mount point is inaccessible.
Made mounting performed here instead an out call.
Streamlined the codebase.

THX723 committed:
Update connect-archive.sh

Fixed typo.

THX723 committed:
Update connect-archive.sh

Added exit codes.

THX723 committed:
Update archiveloop

Rephrased log message for accuracy.

THX723 commented:
@marcone There are more streamlines up for grabs, but I think this is a good stopping point for a merge. Please review.

THX723 committed:
Update connect-archive.sh

Reduced code by using function already exists in archiveloop.

THX723 committed:
Update connect-archive.sh

Typo

THX723 committed:
Update disconnect-archive.sh

typo

THX723 committed:
Update archiveloop

No longer need to check if not using cifs.

THX723 commented:
Found several bugs while validating on my Pi install. Should all be addressed now.

marcone commented on run/archiveloop:
```diff
@@ -63,24 +61,25 @@ then
-      ;;
-  esac
-fi
+case "${ARCHIVE_SYSTEM:-none}" in
+  cifs)
+    [ -n "${SHARE_NAME:-}" ] && export ARCHIVE_MOUNT=/mnt/archive
+    [ -n "${MUSIC_SHARE_NAME:-}" ] && export MUSIC_ARCHIVE_MOUNT=/mnt/musicarchive
+    ;;
+  rsync)
+    export ARCHIVE_SERVER="$RSYNC_SERVER"
+    ;;
+  rclone)
+    export ARCHIVE_SERVER="8.8.8.8"
+    ;;
+  none)
+    export ARCHIVE_SERVER=localhost
+    ;;
+  *)
+    log "ARCHIVE_SERVER not set"
+    exit 1
+    ;;
+esac
```
Aside from the conditional exports of (MUSIC_)ARCHIVE_MOUNT, this behaves differently than before. The previous code would fail if ARCHIVE_SERVER was not set but ARCHIVE_SYSTEM equaled 'cifs'. The current code will let that slide.
The previous code would let you override the default ARCHIVE_SERVER for rsync and rclone, the current code doesn't.
I think the old code should be kept, and the conditional exports of  (MUSIC_)ARCHIVE_MOUNT should be separate. Probably also better to make that conditional on the corresponding .img files existing, instead of on whether SHARE_NAME is set.

marcone replied:
> Probably also better to make that conditional on the corresponding .img files existing, instead of on whether SHARE_NAME is set.

Actually, it should probably be conditional on several things, e.g. if using cifs _and_ cam_disk.bin exists _and_ SHARE_NAME is set, then export ARCHIVE_MOUNT. That should also cover cases where someone later removes a disk image by settings its size to 0, but forgets to unset ARCHIVE_SERVER or SHARE_NAME.

THX723 replied:
Done.

marcone commented on run/archiveloop:
```diff
@@ -673,7 +672,7 @@ function archive_clips () {
   if ! /root/bin/connect-archive.sh
   then
-    log "Couldn't connect archive, skipping archive step"
+    log "Server mount(s) is unreachable. Skipping archive and/or music sync steps."
```
I don't like "Server mount(s) is unreachable" because it's not really correct. It's not trying the reach a mount. The mount is local, the server/share is the remote thing it needs to reach in order to do that mount. I think the original "Couldn't connect archive" was better.

THX723 replied:
Restored.

marcone commented on run/cifs_archive/connect-archive.sh:
```diff
@@ -1,28 +1,15 @@
   then
-    log "Ensuring cam archive is mounted..."
-    if ensure_mountpoint_is_mounted_with_retry "$ARCHIVE_MOUNT"
+    if ! ensure_mountpoint_is_mounted "$mount_point"
```
This inner if/then block is basically equivalent to just `ensure_mountpoint_is_mounted "$mount_point"`

THX723 replied:
Done. Simplified and made it mount with retry just in case.

marcone commented on run/cifs_archive/disconnect-archive.sh:
```diff
@@ -1,29 +1,31 @@
   then
-    log "unmounting $ARCHIVE_MOUNT"
-    if ! umount -f -l "$ARCHIVE_MOUNT"
+    if findmnt --mountpoint "$mount_point" > /dev/null
```
I assume you checked that `findmnt` does not hang when the network or archive becomes unreachable?

THX723 replied:
Yes. `findmnt` merely searches for matching mountpoint defined in `/etc/fstab`. It does not actually test the mountpoint itself.

marcone commented on run/cifs_archive/disconnect-archive.sh:
```diff
@@ -1,29 +1,31 @@
-      log "unmount failed"
-    fi
-  fi
+{
+  unmount_if_set "${ARCHIVE_MOUNT:-}"
+  unmount_if_set "${MUSIC_ARCHIVE_MOUNT:-}"
 } &
```
With this refactoring, it might make sense to run the two calls to `unmount_if_set` in the background so that the two unmounts run concurrently instead of sequentially.

marcone replied:
> With this refactoring, it might make sense to run the two calls to `unmount_if_set` in the background so that the two unmounts run concurrently instead of sequentially.

To clarify, I mean like this:
```
unmount_if_set ... &
unmount_if_set ... &
```

THX723 replied:
Done.

THX723 committed:
Update archiveloop

Restored original ARCHIVE_SERVER code, but expanded it to check for CIFS that allows rsync & rclone, if set, to override CIFS.
CIFS validation includes check for corresponding .bin files.

THX723 committed:
Update archiveloop

Restored original log message; made more sense.

THX723 committed:
Update disconnect-archive.sh

Made both unmounts execute independently in the background.

THX723 committed:
Update connect-archive.sh

Simplified code and used call with retry.

THX723 commented:
@marcone All resolved. Ready for another review.

marcone commented on run/cifs_archive/connect-archive.sh:
```diff
@@ -1,28 +1,9 @@
-  fi
+function mount_if_set() {
+  local mount_point=$1
+  [ -n "$mount_point" ] && ensure_mountpoint_is_mounted_with_retry "$mount_point"
```
I think this should be `[ -z "$mount_point" ] || ensure_mountpoint_is_mounted_with_retry "$mount_point"` so that it doesn't fail when mount_point is empty

THX723 replied:
Right.

THX723 committed:
Update connect-archive.sh

THX723 commented:
@marcone waiting for your review.
